File: UIView+Positioning.m
//
//  UIView+Positioning.m
//
//  Created by Shai Mishali on 5/22/13.
//  Copyright (c) 2013 Shai Mishali. All rights reserved.
//

#import "UIView+Positioning.h"

#define SCREEN_SCALE                    ([[UIScreen mainScreen] scale])
#define PIXEL_INTEGRAL(pointValue)      (round(pointValue * SCREEN_SCALE) / SCREEN_SCALE)

@implementation UIView (Positioning)
@dynamic x, y, width, height, origin, size;

// Setters
-(void)setX:(CGFloat)x{
    CGRect r        = self.frame;
    r.origin.x      = PIXEL_INTEGRAL(x);
    self.frame      = r;
}

-(void)setY:(CGFloat)y{
    CGRect r        = self.frame;
    r.origin.y      = PIXEL_INTEGRAL(y);
    self.frame      = r;
}

-(void)setWidth:(CGFloat)width{
    CGRect r        = self.frame;
    r.size.width    = PIXEL_INTEGRAL(width);
    self.frame      = r;
}

-(void)setHeight:(CGFloat)height{
    CGRect r        = self.frame;
    r.size.height   = PIXEL_INTEGRAL(height);
    self.frame      = r;
}

-(void)setOrigin:(CGPoint)origin{
    self.x          = origin.x;
    self.y          = origin.y;
}

-(void)setSize:(CGSize)size{
    self.width      = size.width;
    self.height     = size.height;
}

-(void)setRight:(CGFloat)right {
    self.x = right - self.width;
}

-(void)setBottom:(CGFloat)bottom {
    self.y = bottom - self.height;
}

-(void)setCenterX:(CGFloat)centerX {
    self.center = CGPointMake(centerX, self.center.y);
}

-(void)setCenterY:(CGFloat)centerY {
    self.center = CGPointMake(self.center.x, centerY);
}

// Getters
-(CGFloat)x{
    return self.frame.origin.x;
}

-(CGFloat)y{
    return self.frame.origin.y;
}

-(CGFloat)width{
    return self.frame.size.width;
}

-(CGFloat)height{
    return self.frame.size.height;
}

-(CGPoint)origin{
    return CGPointMake(self.x, self.y);
}

-(CGSize)size{
    return CGSizeMake(self.width, self.height);
}

-(CGFloat)right {
    return self.frame.origin.x + self.frame.size.width;
}

-(CGFloat)bottom {
    return self.frame.origin.y + self.frame.size.height;
}

-(CGFloat)centerX {
    return self.center.x;
}

-(CGFloat)centerY {
    return self.center.y;
}

-(UIView *)lastSubviewOnX{
    if(self.subviews.count > 0){
        UIView *outView = self.subviews[0];

        for(UIView *v in self.subviews)
            if(v.x > outView.x)
                outView = v;

        return outView;
    }

    return nil;
}

-(UIView *)lastSubviewOnY{
    if(self.subviews.count > 0){
        UIView *outView = self.subviews[0];

        for(UIView *v in self.subviews)
            if(v.y > outView.y)
                outView = v;

        return outView;
    }

    return nil;
}

// Methods
-(void)centerToParent{
    if(self.superview){
        switch ([UIApplication sharedApplication].statusBarOrientation){
            case UIInterfaceOrientationLandscapeLeft:
            case UIInterfaceOrientationLandscapeRight:{
                self.x  =   PIXEL_INTEGRAL((self.superview.height / 2.0) - (self.width / 2.0));
                self.y  =   PIXEL_INTEGRAL((self.superview.width / 2.0) - (self.height / 2.0));
                break;
            }
            case UIInterfaceOrientationPortrait:
            case UIInterfaceOrientationPortraitUpsideDown:{
                self.x  =   PIXEL_INTEGRAL((self.superview.width / 2.0) - (self.width / 2.0));
                self.y  =   PIXEL_INTEGRAL((self.superview.height / 2.0) - (self.height / 2.0));
                break;
            }
        }
    }
}

@end

File: UIView+Positioning.swift
//
//  UIView+Positioning.swift
//
//  Created by Shai Mishali on 1/19/15.
//  Copyright (c) 2015 Shai Mishali. All rights reserved.
//  Updated to Swift 3.0 by Muhammad Assar <abu.assar@gmail.com> on 12/1/2016

import UIKit

extension UIView {
    // MARK: - Basic Properties
    
    /// X Axis value of UIView.
    var x: CGFloat {
        set { self.frame = CGRect(x: newValue,y: self.y,width: self.width,height: self.height) }
        get { return self.frame.origin.x }
    }
    
    /// Y Axis value of UIView.
    var y: CGFloat {
        set { self.frame = CGRect(x: self.x,y: newValue,width: self.width,height: self.height) }
        get { return self.frame.origin.y }
    }
    
    /// Width of view.
    var width: CGFloat {
        set { self.frame = CGRect(x: self.x,y: self.y,width: newValue,height: self.height) }
        get { return self.frame.size.width }
    }
    
    /// Height of view.
    var height: CGFloat {
        set { self.frame = CGRect(x: self.x,y: self.y,width: self.width,height: newValue) }
        get { return self.frame.size.height }
    }
    
    // MARK: - Origin and Size
    
    /// View's origin point.
    var origin: CGPoint {
        set { self.frame = CGRect(x: newValue.x,y: newValue.y,width: self.width,height: self.height) }
        get { return self.frame.origin }
    }
    
    /// View's size.
    var size: CGSize {
        set { self.frame = CGRect(x: self.x,y: self.y,width: newValue.width,height: newValue.height) }
        get { return self.frame.size }
    }
    
    // MARK: - Extra Properties
    
    /// View's right side (x + width).
    var right: CGFloat {
        set { self.x = newValue - self.width }
        get { return self.x + self.width }
    }
    
    /// View's bottom (y + height).
    var bottom: CGFloat {
        set { self.y = newValue - self.height }
        get { return self.y + self.height }
    }
    
    /// View's top (y).
    var top: CGFloat {
        set { self.y = newValue }
        get { return self.y }
    }
    
    /// View's left side (x).
    var left: CGFloat {
        set { self.x = newValue }
        get { return self.x }
    }
    
    /// View's center X value (center.x).
    var centerX: CGFloat {
        set { self.center = CGPoint(x: newValue, y: self.centerY) }
        get { return self.center.x }
    }
    
    /// View's center Y value (center.y).
    var centerY: CGFloat {
        set { self.center = CGPoint(x: self.centerX, y: newValue) }
        get { return self.center.y }
    }
    
    /// Last subview on X Axis.
    var lastSubviewOnX:UIView? {
        get {
            var outView:UIView = self.subviews[0] as UIView
            
            for v in self.subviews as [UIView] {
                if(v.x > outView.x){ outView = v }
            }
            
            return outView
        }
    }
    
    /// Last subview on Y Axis.
    var lastSubviewOnY:UIView? {
        get {
            var outView:UIView = self.subviews[0] as UIView
            
            for v in self.subviews as [UIView] {
                if(v.y > outView.y){ outView = v }
            }
            
            return outView
        }
    }
    
    // MARK: - Bounds Methods
    
    /// X value of bounds (bounds.origin.x).
    var boundsX:CGFloat {
        set{ self.bounds = CGRect(x: newValue, y: self.boundsY, width: self.boundsWidth, height: self.boundsHeight) }
        get{ return self.bounds.origin.x }
    }
    
    /// Y value of bounds (bounds.origin.y).
    var boundsY:CGFloat {
        set { self.frame = CGRect(x: self.boundsX, y: newValue, width: self.boundsWidth, height: self.boundsHeight) }
        get { return self.bounds.origin.y }
    }
    
    /// Width of bounds (bounds.size.width).
    var boundsWidth: CGFloat {
        set { self.frame = CGRect(x: self.boundsX, y: self.boundsY, width: newValue, height: self.boundsHeight) }
        get { return self.bounds.size.width }
    }
    
    /// Height of bounds (bounds.size.height).
    var boundsHeight: CGFloat {
        set { self.frame = CGRect(x: self.boundsX, y: self.boundsY, width: self.boundsWidth, height: newValue) }
        get { return self.bounds.size.height }
    }
    
    // MARK: - Useful Methods
    
    /// Center view to it's parent view.
    func centerToParent(){
        if(self.superview != nil){
            switch(UIApplication.shared.statusBarOrientation){
                case .landscapeLeft:
                    fallthrough
                case .landscapeRight:
                    self.origin = CGPoint(x: (self.superview!.height / 2) - (self.width / 2),
                                          y: (self.superview!.width / 2) - (self.height / 2))
                case .portrait:
                    fallthrough
                case .portraitUpsideDown:
                    self.origin = CGPoint(x: (self.superview!.width / 2) - (self.width / 2),
                                          y: (self.superview!.height / 2) - (self.height / 2))
                case .unknown:
                    return
            }
        }
    }
    
    // MARK: - Private Methods
    private func _pixelIntegral(pointValue:CGFloat) -> CGFloat{
        let scale   = UIScreen.main.scale
        return (round(pointValue * scale) / scale)
    }
}
